File: fwc_edu/fwc_education/report/qsc_top_20_report/qsc_top_20_report.py
# Copyright (c) 2022, Sione Taumoepeau and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
from itertools import groupby
import sys
import datetime
import pandas as pd
import xlsxwriter
import openpyxl 
import numpy as np

from openpyxl import load_workbook
from frappe import _, msgprint, utils
from datetime import datetime, timedelta
from frappe.utils import flt, getdate, datetime, comma_and
from collections import defaultdict
from werkzeug.wrappers import Response
import frappe, erpnext
from collections import defaultdict


def execute(filters=None):
	if not filters: filters = {}
	columns, data = [], []

	formLevel = filters.get('level')
	term = filters.get('academic_term')
	
	program = get_program(formLevel)

	if formLevel in ("L1", "L2", "L3", "L4"):
		
#		frappe.msgprint(_("Level {0}").format(program))

		midyear_40 = frappe.db.sql("""SELECT tabAR.student,tabAR.student_name,
					ROUND(((SUM(tabAR.total_score)/800*100)*40/100), 2) AS 'MidYear_Score'
					FROM `tabAssessment Result` as tabAR
					WHERE tabAR.docstatus = 1
					AND tabAR.not_included = 0
					AND tabAR.program LIKE %s
					AND tabAR.academic_term = "2022 (Term 1)"
					GROUP BY tabAR.student""", ("%%%s%%" % program), as_dict=1)
	
		final_60 = frappe.db.sql("""SELECT tabAR.student,tabAR.student_name,
					ROUND(((SUM(tabAR.total_score)/800*100)*60/100), 2) AS 'Total_Score'
					FROM `tabAssessment Result` as tabAR
					WHERE tabAR.docstatus = 1
					AND tabAR.not_included = 0
					AND tabAR.program LIKE %s
					AND tabAR.academic_term = %s
					GROUP BY tabAR.student""", ("%%%s%%" % program, term), as_dict=1)
	
	if formLevel in ("L5", "L6", "L7"):
#		frappe.msgprint(_("Level 2 {0}").format(program))

		midyear_40 = frappe.db.sql("""SELECT tabAR.student,tabAR.student_name,
					ROUND(((SUM(tabAR.total_score)/600*100)*40/100), 2) AS 'MidYear_Score'
					FROM `tabAssessment Result` as tabAR
					WHERE tabAR.docstatus = 1
					AND tabAR.not_included = 0
					AND tabAR.program LIKE %s
					AND tabAR.academic_term = "2022 (Term 1)"
					GROUP BY tabAR.student""", ("%%%s%%" % program), as_dict=1)
	
		final_60 = frappe.db.sql("""SELECT tabAR.student,tabAR.student_name,
					ROUND(((SUM(tabAR.total_score)/600*100)*60/100), 2) AS 'Total_Score'
					FROM `tabAssessment Result` as tabAR
					WHERE tabAR.docstatus = 1
					AND tabAR.not_included = 0
					AND tabAR.program LIKE %s
					AND tabAR.academic_term = %s
					GROUP BY tabAR.student""", ("%%%s%%" % program, term), as_dict=1)


#	df_40 = pd.DataFrame.from_records(midyear_40)
#	df_60 = pd.DataFrame.from_records(final_60)
	

	dataframe = pd.DataFrame.from_records(midyear_40)
	df_total = pd.DataFrame.from_records(final_60)

	df_total = df_total.pivot_table(index=('student_name'), values='Total_Score')
#	lessons = dataframe.course.unique().tolist()
	dataframe = dataframe.pivot_table(index=('student_name'), values=('MidYear_Score'))

	

	dataframe['Overall'] = df_total['Total_Score'] + dataframe['MidYear_Score']

	dataframe['Comments'] = " "
	
	dataframe['FinalSecond'] = dataframe['Overall'] - dataframe['MidYear_Score']

	dataframe['FinalSecond'] = round(dataframe['FinalSecond'], 2)

	dataframe['Overall'] = round(dataframe['Overall'] ,2)
	dataframe.fillna(0, inplace = True)

	dataframe = dataframe.sort_values(by="Overall", ascending=False)
	
	#dataframe = dataframe.iloc[:20]

	columns = [ { "fieldname": "rank", "label": _("Position"), "fieldtype": "Data", "width": 200 }]
	columns = [ { "fieldname": "student_name", "label": _("Student"), "fieldtype": "Data", "width": 200 }]
	columns+=[ { "fieldname": "Comments", "label": _("Kolo e Ui me Ai"), "fieldtype": "Data", "width": 500}]

	columns+=[ { "fieldname": "MidYear_Score", "label": _("Mid Year 40%"), "fieldtype": "Float", "width": 100 }]

	columns+=[ { "fieldname": "FinalSecond", "label": _("Second Half 60%"), "fieldtype": "Float", "width": 100 }]

	columns+=[ { "fieldname": "Overall", "label": _("Overall"), "fieldtype": "Float", "width": 100 }]

	data = dataframe.reset_index().to_dict('records')
	
	return columns, data

def get_program(level):
	
	if level == "L1":
		return "Form 1"
	if level == "L2":
		return "Form 2"
	if level == "L3":
		return "Form 3"
	if level == "L4":
		return "Form 4"
	if level == "L5":
		return "Form 5"
	if level == "L6":
		return "Form 6"
	if level == "L7":
		return "Form 7"
	

File: fwc_edu/fwc_education/report/qsc_top_20_report/qsc_top_20_report.js
// Copyright (c) 2022, Sione Taumoepeau and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["QSC TOP 20 Report"] = {
	filters: [
		{
			fieldname:"program",
			label: __("Program"),
			fieldtype: "Link",
			options : 'Program',
			width: "100px",
			reqd: 0,
			hidden:1
		},
		{
			"fieldname":"academic_term",
			"label":__("Academic Term"),
			"fieldtype":"Link",
			"options": "Academic Term",
			"width": "90px",
			"reqd": 1
		},
		{
			fieldname:"level",
			label: __("Choose Level"),
			fieldtype: "Select",
			options : [
				"",
				{ "value": "L1", "label": __("FORM 1 LEVEL") },
				{ "value": "L2", "label": __("FORM 2 LEVEL") },
				{ "value": "L3", "label": __("FORM 3 LEVEL") },
				{ "value": "L4", "label": __("FORM 4 LEVEL") },
				{ "value": "L5", "label": __("FORM 5 LEVEL") },
				{ "value": "L6", "label": __("FORM 6 LEVEL") },
				{ "value": "L7", "label": __("FORM 7 LEVEL") },
			],
			reqd: 1,

			on_change: function() {
				var levelClass = ""
				var level = frappe.query_report.get_filter_value('level');
				if (level){
					if (level = 'L1'){
						levelClass = "Form 1"
					}else if (level = 'L2'){
						levelClass = "Form 2"
					}else if (level = 'L3'){
						levelClass = "Form 3"
					}else if (level = 'L4'){
						levelClass = "Form 4"
					}else if (level = 'L5'){
						levelClass = "Form 5"
					}else if (level = 'L6'){
						levelClass = "Form 6"
					}else if (level = 'L7'){
						levelClass = "Form 7"
					}
					frappe.query_report.set_filter_value('classLevel', levelClass)

				} else {
					frappe.query_report.set_filter_value('classLevel', "");
				}
			} 
		},
		{
			fieldname:"classLevel",
			label: __("Class Level"),
			fieldtype: "Data",
			reqd: 0,
			hidden:1
		},
	]
};
